### PR TITLE
[BEAM-13051][C] Enable pylint warnings (misplaced-bare-raise/no-member)

### PR DESCRIPTION
**Pull request by roger-mike**
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: sdks/python/apache_beam/dataframe/frames.py [hunk 1]
```diff
@@ -3863,7 +3863,6 @@ def idxmin(self, **kwargs):
   def idxmax(self, **kwargs):
     return self._idxmaxmin_helper('idxmax', **kwargs)
 
-
```
Comment (roger-mike) [inline sdks/python/apache_beam/dataframe/frames.py:3866]: Unrelated change

File: sdks/python/apache_beam/io/external/xlang_kinesisio_it_test.py [hunk 2]
```diff
@@ -290,7 +290,7 @@ def get_first_shard_id(self, stream_name):
       time.sleep(2)
       if i == retries - 1:
         logging.error('Could not initialize kinesis stream')
-        raise
+        raise RuntimeError()
```
Comment (roger-mike) [inline sdks/python/apache_beam/io/external/xlang_kinesisio_it_test.py:293]: Since there's no active exception, this bare raise raises a RuntimeError. I made it explicit here.

File: sdks/python/apache_beam/runners/common.py [hunk 3]
```diff
@@ -1262,7 +1262,7 @@ def finalize(self):
 
   def _reraise_augmented(self, exn):
     if getattr(exn, '_tagged_with_step', False) or not self.step_name:
-      raise
+      raise exn
```
Comment (roger-mike) [inline sdks/python/apache_beam/runners/common.py:1265]: Just explicitly reraised the active exception.

File: sdks/python/apache_beam/io/external/xlang_kinesisio_it_test.py [hunk 4]
```diff
@@ -267,10 +267,10 @@ def create_stream(self, stream_name):
         )
         time.sleep(2)
         break
-      except:  # pylint: disable=bare-except
+      except Exception as e:
         if i == retries - 1:
           logging.error('Could not create kinesis stream')
-          raise
+          raise e
```
Comment (roger-mike) [inline sdks/python/apache_beam/io/external/xlang_kinesisio_it_test.py:273]: Raised the active exception at this point.

File: sdks/python/apache_beam/runners/portability/portable_runner.py [hunk 5]
```diff
@@ -628,12 +628,10 @@ def _cleanup(self, on_exit=False):
...
         callback()
-      except Exception:
-        has_exception = True
-    self._cleanup_callbacks = ()
-    if has_exception:
-      raise
+      except Exception as e:
+        raise e
+      finally:
+        self._cleanup_callbacks = ()
```
Comment (roger-mike) [inline sdks/python/apache_beam/runners/portability/portable_runner.py:637]: `self._cleanup_callbacks = ()` executes whether there's an exception or not.

**Comment by roger-mike** [comment]
The `no-member` warning remained enabled because most of the warnings are false positives related to the message classes generated by protobuf or because didn't find the class inside the `apache_beam` module (maybe is related to the `import-error` warning)

**Comment by roger-mike** [comment]
R: @TheNeuralBit could you take a look? Thanks.

**Comment by roger-mike** [comment]
Also, there are `no-member` warnings in test classes (`cache_manager_test.py` and `avroio_test.py`). Some of them happen because they use methods/variables that are not yet inherited. For example:
```
class BaseTest(object):
  def test_a(self):
    self.a = 1  # throws no-member on a
    self.assertTrue(...)  # throws no-member on assertTrue

class TestA(BaseTest, unittest.TestCase):
  def __init__(self):
    self.variable_a = 0

  def test_b(self):
    self.a = 1
    self.assertTrue(...)
```
These could be fixed, but still not sure how to proceed with the rest of the `no-member` warnings.

**Comment by roger-mike** [comment]
R: @pabloem could you take a look? Thanks.

File: sdks/python/apache_beam/io/external/xlang_kinesisio_it_test.py (hunk 2 above)
Comment (pabloem) [reply sdks/python/apache_beam/io/external/xlang_kinesisio_it_test.py:293]:
```suggestion
        raise RuntimeError("Unable to initialize Kinesis Stream %s. Status: %s",
            stream['StreamDescription']['StreamName'], stream['StreamDescription']['StreamStatus'])
```

File: sdks/python/apache_beam/runners/portability/portable_runner.py (hunk 5 above)
Comment (pabloem) [reply sdks/python/apache_beam/runners/portability/portable_runner.py:637]: this code has slightly different behavior, right? The new code may not call all of the cleanup callbacks. If one of them errors out before the end, then the final callbacks will not be processed, right?

Maybe you can store all exceptions, and raise a RuntimeException with information about all of them?

File: sdks/python/apache_beam/runners/common.py (hunk 3 above)
Comment (pabloem) [reply sdks/python/apache_beam/runners/common.py:1265]: lgtm

File: sdks/python/apache_beam/io/external/xlang_kinesisio_it_test.py (hunk 4 above)
Comment (pabloem) [reply sdks/python/apache_beam/io/external/xlang_kinesisio_it_test.py:273]: lgtm

**Review by pabloem — COMMENTED** [review]
added two comments. other than that, this LGTM

File: sdks/python/apache_beam/runners/portability/portable_runner.py (hunk 5 above)
Comment (roger-mike) [reply sdks/python/apache_beam/runners/portability/portable_runner.py]: Done 👍 , not sure if the format is correct but it does what you suggest.

**Comment by pabloem** [comment]
I'll ask @AnandInguva to take a look at this. Thanks @AnandInguva

**Comment by roger-mike** [comment]
@pabloem What do you think of the `no-member` warnings? Do you think some of them could be fixed? Most of them are false positives related to the `apache-beam` module, I think these can remain skipped.

File: sdks/python/apache_beam/runners/portability/portable_runner.py [hunk 6]
```diff
@@ -628,12 +628,17 @@ def _cleanup(self, on_exit=False):
...
+        callback_exceptions.append(e)
+
     self._cleanup_callbacks = ()
-    if has_exception:
-      raise
+    if callback_exceptions:
+      formatted_exceptions = ''.join(
+          [f"\n\t{repr(e)}" for e in callback_exceptions])
+      raise RuntimeError(
+          'Cleanup callbacks raised the following errors: {}'.format(
```
Comment (AnandInguva) [inline sdks/python/apache_beam/runners/portability/portable_runner.py:643]: May be we just mention errors instead of `Cleanup callbacks` as  `_cleanup` is an internal method and may be it creates confusion among users who tries to look at the logs.

File: sdks/python/apache_beam/runners/portability/portable_runner.py [hunk 7]
```diff
@@ -628,12 +628,15 @@ def _cleanup(self, on_exit=False):
...
+      except Exception as e:
+        callback_exceptions.append(e)
+
     self._cleanup_callbacks = ()
-    if has_exception:
-      raise
+    if callback_exceptions:
+      formatted_exceptions = ''.join(
+          [f"\n\t{repr(e)}" for e in callback_exceptions])
+      raise RuntimeError('Errors: {}'.format(formatted_exceptions))
```
Comment (roger-mike) [inline sdks/python/apache_beam/runners/portability/portable_runner.py:642]: @AnandInguva is something like this ok?

Comment (AnandInguva) [reply sdks/python/apache_beam/runners/portability/portable_runner.py:642]: LGTM

**Review by AnandInguva — APPROVED** [review]

**Comment by TheNeuralBit** [comment]
Is anything blocking this @pabloem?

**Comment by pabloem** [comment]
no! I'm happy to merge

**Comment by pabloem** [comment]
Run PythonLint PreCommit

**Comment by roger-mike** [comment]
Run PythonLint PreCommit
